packages/vite/src/node/config.ts

```diff
@@ -205,18 +213,37 @@ export interface DevEnvironmentOptions {
   // fs: { strict?: boolean, allow, deny }
 }
 
-export type ResolvedDevEnvironmentOptions = Required<
-  Omit<DevEnvironmentOptions, 'createEnvironment'>
-> & {
-  // TODO: Should we set the default at config time? For now, it is defined on server init
-  createEnvironment:
-    | ((
-        name: string,
-        config: ResolvedConfig,
-      ) => Promise<DevEnvironment> | DevEnvironment)
-    | undefined
+function createDefaultClientDevEnvironment(
+  name: string,
+  config: ResolvedConfig,
+  context: CreateDevEnvironmentContext,
+) {
+  return new DevEnvironment(name, config, {
+    hot: context.ws,
+  })
+}
+
+function createDefaultSsrDevEnvironment(
+  name: string,
+  config: ResolvedConfig,
+): DevEnvironment {
+  return createNodeDevEnvironment(name, config, {
+    hot: createServerHotChannel(),
+  })
```

Comment:
@sheremet-va before, we only used this for the `ssr` environment. If I recall correctly, this setup was only needed to be backward compatible with the current `ssr` environment, no? In that case, what should be default `createEnvironment` for other SSR-like environments?

Reply:
Feels weird to me that it needs a default, but I guess it's just a dev environment? Why does it need anything special?

Reply:
We need a default `createEnvironment` or we need to make `createEnvironment` required for any custom environment that is not named `'client'` and `'ssr'`.

It doesn't need anything special, we only need what `hot` should be by default in that case. Before this PR, by default it was `hot: ws`:
```js
new DevEnvironment(name, config, { hot: ws })
```
Right now, it is still that if `consumer: 'client'`, and if `consumer: 'server'`, it is using the same as in the default `ssr` client.
```js
createNodeDevEnvironment(name, config, {
  hot: createServerHotChannel(),
})
```
We should only use this for `'ssr'`, no?

Reply:
So the default for `consumer: 'client'` (for example, someone creates a `'electron'` environment), should be:
```js
new DevEnvironment(name, config, { hot: ws })
```
and the default for `consumer: 'server'` (for example, someone creates a `'rsc'` environment), should be:
```js
new DevEnvironment(name, config, { hot: false })
```

Or should the default be `{ hot: false }` for every environment that is not named `'client'` and `'ssr'`?

Reply:
I think it should be `false` for every unknown environment (so, not a `client` or `ssr`). HMR is implemented by the consumer - we cannot assume what they want to use to communicate with the runner, so they will always create it if needed. We initialize HMR only for backwards compatibility and because we know how these environments communicate

Reply:
Makes sense. Changed to `hot: false` by default at [chore: default dev environment has hot: false](https://github.com/vitejs/vite/pull/17791/commits/855c54943fb3cedf84febbcf396a8fb1984f3176)